pyodide_build/recipe/graph_builder.py

```diff
@@ -647,26 +648,50 @@ def run(self, n_jobs: int, already_built: set[str]) -> None:
                         self.build_queue.put((job_priority(dependent), dependent))
 
 
+def _run(cmd, *args, check=False, **kwargs):
+    result = subprocess.run(cmd, *args, **kwargs, check=check)
+    if result.returncode != 0:
+        logger.error("ERROR: command failed %s", " ".join(cmd))
+        exit_with_stdio(result)
+    return result
+
+
 def _ensure_rust_toolchain():
     rust_toolchain = build_env.get_build_flag("RUST_TOOLCHAIN")
-    result = subprocess.run(
-        ["rustup", "toolchain", "install", rust_toolchain], check=False
-    )
-    if result.returncode == 0:
-        result = subprocess.run(
+    _run(["rustup", "toolchain", "install", rust_toolchain])
+    _run(["rustup", "default", rust_toolchain])
+
+    url = build_env.get_build_flag("RUST_EMSCRIPTEN_TARGET_URL")
+    if not url:
+        # Install target with rustup target add
+        _run(
             [
                 "rustup",
                 "target",
                 "add",
                 "wasm32-unknown-emscripten",
                 "--toolchain",
                 rust_toolchain,
-            ],
-            check=False,
+            ]
         )
-    if result.returncode != 0:
-        logger.error("ERROR: rustup toolchain install failed")
-        exit_with_stdio(result)
+        return
+
+    # Install target from url
+    result = _run(
+        ["rustup", "which", "--toolchain", rust_toolchain, "rustc"],
+        capture_output=True,
+        text=True,
+    )
+    toolchain_root = Path(result.stdout).parents[1]
+    rustlib = toolchain_root / "lib/rustlib"
+    install_token = rustlib / "wasm32-unknown-emscripten_install-url.txt"
```

Comment:
Is this file something you added manually in your custom toolchain? If so, let's add a comment about it.